opacus/layers/dp_lstm.py

```diff
@@ -114,29 +175,60 @@ def forward(
             - ``c_n`` is of shape ``[B, H]`` tensor containing the cell state for ``t = T``.
         """
 
-        seq_length, batch_sz, _ = x.shape
-        if self.reverse:
-            x = x.flip(0)
-        x = torch.unbind(x, dim=0)
+        if batch_sizes is not None:
+            seq_length = batch_sizes.size(0)
+            if self.reverse:
+                x = tuple(reversed(x))
+                batch_sizes = batch_sizes.flip(0)
+        else:
+            seq_length, batch_sz, _ = x.shape
+            if self.reverse:
+                x = x.flip(0)
+            x = torch.unbind(x, dim=0)
 
         h_0, c_0 = state_init
 
         h_n = [h_0]
         c_n = [c_0]
+        batch_size_prev = h_0.shape[0]
 
         for t in range(seq_length):
-            h_next, c_next = self.cell(x[t], h_n[t], c_n[t])
+            if batch_sizes is not None:
+                batch_size_t = batch_sizes[t].item()
+                delta = batch_size_t - batch_size_prev
+                if delta > 0:
+                    h_cat = torch.cat((h_n[t], h_0[batch_size_prev:batch_size_t, :]), 0)
+                    c_cat = torch.cat((c_n[t], c_0[batch_size_prev:batch_size_t, :]), 0)
+                    h_next, c_next = self.cell(
+                        x[t], h_cat, c_cat, batch_size_t, self.reverse
+                    )
+                else:
+                    h_next, c_next = self.cell(
+                        x[t], h_n[t], c_n[t], batch_size_t, self.reverse
+                    )
+            else:
+                h_next, c_next = self.cell(x[t], h_n[t], c_n[t])
             if self.dropout:
                 h_next = self.dropout_layer(h_next)
             h_n.append(h_next)
             c_n.append(c_next)
+            batch_size_prev = h_next.shape[0]
 
-        h_n = torch.stack(h_n[1:], dim=0)  # [T, B, H], init step not part of output
+        if batch_sizes is None:
+            h_n = torch.stack(h_n[1:], dim=0)  # [T, B, H], init step not part of output
 
-        return (
-            h_n.flip(0) if self.reverse else h_n,  # Flip the output...
-            (h_n[-1], c_n[-1]),  # ... But not the states
-        )
+            return (
+                h_n.flip(0) if self.reverse else h_n,  # Flip the output...
+                (h_n[-1], c_n[-1]),  # ... But not the states
+            )
+        else:
+            seq_lengths = compute_seq_lengths(batch_sizes)
+            h_temp, c_temp = h_n[1:], c_n[1:]
+            h_last, c_last = compute_last_states(h_temp, c_temp, seq_lengths)
+            if self.reverse:
+                h_temp = tuple(reversed(h_temp))
+
+            return h_temp, (h_last, c_last)
 
 
 class BidirectionalDPLSTMLayer(nn.Module):
```

Comment:
In my opinion, this is the first class where we can have an `if` and switch between the "normal" layers or the ones for PackedSequences